types/mz/child_process.d.ts

```diff
@@ -1,46 +1,44 @@
 // Modified from the node.js definitions.
 // https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/child_process.d.ts
 
-import {
-    ChildProcess,
-    ExecException,
-    ExecFileOptions,
-    ExecFileOptionsWithStringEncoding,
-    ExecOptions,
-} from "child_process";
+import { ChildProcess, ExecException, ExecFileOptions, ExecOptions } from "child_process";
 export * from "child_process";
 
+export interface ExecFileOptionsWithStringEncoding extends ExecFileOptions {
+    encoding: BufferEncoding;
+}
+
 export function exec(
     command: string,
     callback: (error: ExecException | null, stdout: string, stderr: string) => void,
 ): ChildProcess;
 export function exec(
     command: string,
-    options: { encoding: "buffer" | null | undefined } & ExecOptions,
+    options: { encoding: "buffer" | null | undefined } & Omit<ExecOptions, "encoding">,
     callback: (error: ExecException | null, stdout: Buffer, stderr: Buffer) => void,
 ): ChildProcess;
 export function exec(
     command: string,
-    options: ({ encoding?: BufferEncoding | undefined } & ExecOptions) | null | undefined,
+    options: ({ encoding?: BufferEncoding | undefined } & Omit<ExecOptions, "encoding">) | null | undefined,
     callback: (error: ExecException | null, stdout: string, stderr: string) => void,
 ): ChildProcess;
 export function exec(
     command: string,
-    options: ({ encoding?: string | null | undefined } & ExecOptions) | null | undefined,
+    options: ({ encoding?: string | null | undefined } & Omit<ExecOptions, "encoding">) | null | undefined,
     callback: (error: ExecException | null, stdout: string | Buffer, stderr: string | Buffer) => void,
 ): ChildProcess;
 
 export function exec(
     command: string,
-    options: { encoding: "buffer" | null | undefined } & ExecOptions,
+    options: { encoding: "buffer" | null | undefined } & Omit<ExecOptions, "encoding">,
 ): Promise<[Buffer, Buffer]>;
 export function exec(
     command: string,
-    options?: ({ encoding?: BufferEncoding | undefined } & ExecOptions) | null,
+    options?: ({ encoding?: BufferEncoding | undefined } & Omit<ExecOptions, "encoding">) | null,
 ): Promise<[string, string]>;
 export function exec(
     command: string,
-    options?: ({ encoding?: string | null | undefined } & ExecOptions) | null,
+    options?: ({ encoding?: string | null | undefined } & Omit<ExecOptions, "encoding">) | null,
```

Comment:
These are now incorrect, though? An encoding of `undefined` can never produce `Buffer` output; an encoding of `null` can never produce `string` output.

I think it would be best to merge the fix for the definitively wrong downstream package, then revert this latest set of changes here.

Reply:
It is what it is, at least at the time these types were created / edited. I believe the original intention is to be as inclusive as possible at the lowest overloads, so that a "default" signature always applied.

But you are right, these types worth to be revisited later. My goal for now is to let test pass and preserve the current behaviour for these types at the moment.

Reply:
Finally got time (and patience) to come back to this PR.

In latest commit I have introduce the method signature change into `mz` & `child-process-promise`. Since these packages are old in history, I basically redefine the affected option types directly into those packages, so that consumers at lower versions (e.g. v16, or even v14) can enjoy a non-breaking type upgrade at their will.

See if this latest push can solve for what you suggest. :)